sway-lsp/src/server.rs

```diff
@@ -62,23 +63,26 @@ impl Backend {
     }
 
     async fn parse_project(&self, uri: Url, workspace_uri: Url, session: Arc<Session>) {
-        // pass in the temp Url into parse_project, we can now get the updated AST's back.
-        let diagnostics = match session.parse_project(&uri) {
-            Ok(diagnostics) => diagnostics,
-            Err(err) => {
-                tracing::error!("{}", err.to_string().as_str());
-                if let LanguageServerError::FailedToParse { diagnostics } = err {
-                    diagnostics
-                } else {
-                    Diagnostics {
-                        warnings: vec![],
-                        errors: vec![],
-                    }
+        let uri_clone = uri.clone();
+        let session_clone = session.clone();
+        // Run parse_project in a blocking thread, because parsing is not async.
+        let should_publish = task::spawn_blocking(move || {
+            // Pass in the temp Url into parse_project, we can now get the updated AST's back.
+            match session_clone.parse_project(&uri_clone) {
+                Ok(should_publish) => should_publish,
+                Err(err) => {
+                    tracing::error!("{}", err.to_string().as_str());
+                    matches!(err, LanguageServerError::FailedToParse)
                 }
             }
-        };
-        self.publish_diagnostics(&uri, &workspace_uri, session, diagnostics)
-            .await;
+        })
+        .await
+        .unwrap_or_default();
+
+        if should_publish {
+            self.publish_diagnostics(&uri, &workspace_uri, session)
+                .await;
+        }
```

Comment:
Perhaps something like this is cleaner?

```rust
async fn parse_project(&self, uri: Url, workspace_uri: Url, session: Arc<Session>) {
    let should_publish = run_blocking_parse_project(uri.clone(), session.clone()).await;
    if should_publish {
        self.publish_diagnostics(&uri, &workspace_uri, session).await;
    }
}
```

and then have a free function declared outside of `Backend`

```rust
async fn run_blocking_parse_project(uri: Url, session: Arc<Session>) -> bool {
    task::spawn_blocking(move || {
        match session.parse_project(&uri) {
            Ok(should_publish) => should_publish,
            Err(err) => {
                tracing::error!("{}", err);
                matches!(err, LanguageServerError::FailedToParse)
            }
        }
    })
    .await
    .unwrap_or_default()
}
```